core/dbt/include/global_project/macros/materializations/models/table/columns_spec_ddl.sql

```diff
@@ -8,8 +8,22 @@
 
 {% macro table_columns_and_constraints() %}
   {# loop through user_provided_columns to create DDL with data types and constraints #}
+    {%- if defer_relation and config.get('contract').inferred -%}
+      {% set prod_columns = get_column_schema_from_query(get_empty_subquery_sql("select * from " ~ defer_relation)) %}
+      {% set raw_columns = {} %}
+      {% for column in prod_columns -%}
+        {# TODO: reconsider this mutating operation, could be done with an additional warehouse call #}
```

Comment:
👍 At minimum, we should probably be deepcopying to avoid mutating the user-provided value stored in the context... right?

Reply:
We can either:
- plumb it through (but this takes some time, and tightly couples the methods)
- make the warehouse call twice (and consider caching/optimization possibilities in the future)